### PR TITLE
Goa 2124 g protein set endpoint kruk etc

### PR DESCRIPTION
**Pull request by TonyWardell**
Add target set as  searchable, non-retrievable list of strings to the Gene Product Model.
The target set list is populated at gene product indexing time.
An endpoint in the gene product controller allows looking up the list of gene products for a target set.

### CODE REVIEW

File: geneproduct-common/src/main/java/uk/ac/ebi/quickgo/geneproduct/common/document/GeneProductDocument.java [hunk 1]
```diff
@@ -54,6 +54,9 @@
     @Field(GeneProductFields.PARENT_ID)
     public String parentId;
 
+    @Field(GeneProductFields.TARGET_SET)
+    public List<String> targetSet;
```
Comment (TonyWardell) [inline geneproduct-common/src/main/java/uk/ac/ebi/quickgo/geneproduct/common/document/GeneProductDocument.java]: Done.

File: geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerIT.java [hunk 2]
```diff
@@ -138,6 +141,29 @@ public void finds200IfNoResultsBecauseIdsDoNotExist() throws Exception {
...
+                .andExpect(status().isOk());
+
+        int index = 0;
+        for (String id : validIdList) {
+            expectFields(result, id, "$.results[" + index++ + "].");
+        }
+    }
+
+    @Test
+    public void targetSetLookUpUsingInvalidValueReturnsEmptyResults() throws Exception {
```
Comment (TonyWardell) [inline geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerIT.java]: Done.
Added this check to finds200IfNoResultsBecauseIdsDoNotExist also.

File: geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerIT.java [hunk 3]
```diff
@@ -138,6 +141,29 @@ public void finds200IfNoResultsBecauseIdsDoNotExist() throws Exception {
...
+        ResultActions result = mockMvc.perform(get(buildGeneProductTargetSetURL(NON_EXISTENT_TARGET_SET_NAME)));
+        result.andDo(print())
+              .andExpect(status().isOk());
+    }
+
```
Comment (TonyWardell) [inline geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerIT.java]: Then the value 'targetset' is passed to findById() in the controller....

File: indexing/src/test/java/uk/ac/ebi/quickgo/index/geneproduct/GeneProductDocumentConverterTest.java [hunk 4]
```diff
@@ -144,6 +145,18 @@ public void converts3SynonymsInGeneProductToListWith3Synonyms() throws Exception
     }
 
     @Test
+    public void converts3TargetSetValuesInPropertiesToListWith3TargetSets() throws Exception {
+        List<String> targetSets = Arrays.asList("KRUK", "Parkinsons", "BHF-UCL");
+
+        String targetSet = concatProperty(TARGET_SET_KEY, concatStrings(targetSets, SPECIFIC_VALUE_DELIMITER));
+        geneProduct.properties = concatStrings(Arrays.asList(targetSet), INTER_VALUE_DELIMITER);
```
Comment (TonyWardell) [inline indexing/src/test/java/uk/ac/ebi/quickgo/index/geneproduct/GeneProductDocumentConverterTest.java]: Done.

File: geneproduct-rest/src/main/java/uk/ac/ebi/quickgo/geneproduct/service/GeneProductService.java [hunk 5]
```diff
@@ -20,4 +20,11 @@
 	 * chosen information
 	 */
 	List<GeneProduct> findById(List<String> ids);
+
+	/**
+	 * Find gene products core data for all gene products associated with the named target set.
+	 * @param name
```
Comment (TonyWardell) [inline geneproduct-rest/src/main/java/uk/ac/ebi/quickgo/geneproduct/service/GeneProductService.java]: Fixed.

File: geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerIT.java [hunk 6]
```diff
@@ -161,7 +163,17 @@ public void targetSetLookUpUsingValidValueReturnsMultiGeneProduct() throws Excep
...
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.numberOfHits").value(0))
+                .andExpect(jsonPath("$.results").isArray());
+    }
+
+    @Test
+    public void targetSetLookUpUsingEmptyStringReturnsBadRequest() throws Exception {
+        ResultActions result = mockMvc.perform(get(buildGeneProductTargetSetURL("")));
+        result.andDo(print())
+                .andExpect(jsonPath("$.messages", hasItem(is("Provided ID: 'targetset' is invalid"))))
```
Comment (rantunesboreas) [inline geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerIT.java]: Not quite sure where 'targetset' has come from, could you explain?

File: geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerTest.java [hunk 7]
```diff
@@ -113,13 +113,6 @@ public void retrieveSingleGeneProduct() {
         assertThat(response.getBody().getResults(), hasSize(1));
     }
 
-    @Test
```
Comment (rantunesboreas) [inline geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerTest.java]: You have only removed one of the repeated tests, could you remove the remaning ones?

File: indexing/src/main/java/uk/ac/ebi/quickgo/index/geneproduct/GeneProductDocumentConverter.java [hunk 8]
```diff
@@ -20,6 +20,13 @@
  * @author Ricardo Antunes
  */
 public class GeneProductDocumentConverter implements ItemProcessor<GeneProduct, GeneProductDocument> {
+
+    /**
+     * The following values define how the contents of the Gene Product properties are describes
```
Comment (rantunesboreas) [inline indexing/src/main/java/uk/ac/ebi/quickgo/index/geneproduct/GeneProductDocumentConverter.java]: replace "describes" with "described"

File: geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerIT.java (hunk 6 above)
Comment (TonyWardell) [inline geneproduct-rest/src/test/java/uk/ac/ebi/quickgo/geneproduct/controller/GeneProductControllerIT.java]: MockHttpServletResponse:
           Status = 400
    Error message = null
          Headers = {Content-Type=[application/json]}
     Content type = application/json
             Body = {"url":"http://localhost/geneproduct/targetset/","messages":["Provided ID: 'targetset' is invalid"]}
    Forwarded URL = null
   Redirected URL = null
          Cookies = []

If not targetset value is provided, then the literal targetset is interpreted as a geneproduct id.

File: indexing/src/main/java/uk/ac/ebi/quickgo/index/geneproduct/GeneProductDocumentConverter.java (hunk 8 above)
Comment (TonyWardell) [inline indexing/src/main/java/uk/ac/ebi/quickgo/index/geneproduct/GeneProductDocumentConverter.java]: Done.

**Pull request merged**